src/lib.rs

```diff
@@ -19,288 +23,241 @@ use syntax::ast_map;
-                        visit::walk_expr(self, ex)
-                    }
                 }
+                visit::walk_expr(self, e);
             }
-            ExprMethodCall(ref spanid, ref tys, ref params) => {
-                let method_call = ty::MethodCall::expr(ex.id);
-                let method_map = self.cx.tcx.method_map.borrow();
-                let method = method_map.get(&method_call);
-                if let Some(method_callee) = method {
-                    if let ty::MethodStatic(id) = method_callee.origin {
-                        let attrs = get_attrs_opt(self.cx.tcx, id);
-                        if let Some(v) = attrs {
-                            if v.iter().any(|item| item.check_name("allowed_on_protected")) {
-                                for param in params {
-                                    if let ExprPath(_, _) = param.node {
-                                        // It's an ident within an allowed method call,
-                                        // it's fine!
-                                        self.cx.tcx.sess.span_note(ex.span, "Allowed usage of type. Carry on!");
-                                        return;
-                                    } else {
-                                        // It could be a situation like
-                                        // allowed_fn(foo, bar, {foo(); bar(); unsafe_drop_fn(protected); baz()})
-                                        // this ensures that no trickery happens
-                                        self.visit_expr(&*param);
-                                    }
-                                }
-                            } else if v.iter().any(|item| item.check_name("allowed_drop")) {
-                                for param in params {
-                                    if let ExprPath(_, _) = param.node {
-                                        // It's an ident within an allowed drop call,
-                                        // we should remove it from the map if it was there
-                                        let def = self.cx.tcx.def_map.borrow().get(&param.id).map(|&x| x);
-                                        if let Some(PathResolution { base_def: DefLocal(id), .. }) = def {
-                                            self.cx.tcx.sess.span_note(ex.span, "Properly dropped!");
-                                            self.map.remove(&id);
-                                            return;
-                                        }
-                                    } else {
-                                        self.visit_expr(&*param)
-                                    }
-                                }
-                            }
-                        }
+            ExprCall(_, _) | ExprMethodCall(_, _, _) => {
+                visit::walk_expr(self, e);
+            }
+            ExprAddrOf(_, _) => {
+                // Ignore children
```

Comment:
Perhaps. But it would be easier to simply do what I listed above:

 - If the `expr` inside the `ExprAddrOf` is an `ExprPath`, then it's something simple like `&x`, which we can ignore.
 - If it's more complicated, recurse. Worst case scenario we count a borrow as a move, and emit an additional error. Not a big deal IMO.